first step of constraint transformation

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80147
* #80110
* __->__ #80102
* #80095

Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

migeed-z committed:
first step of constraint transformation

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"




[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"




[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387412).

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. 

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/constraint_transformation.py:
```diff
@@ -0,0 +1,724 @@
+        transformed, counter = transform_apply_broadcasting(constraint, counter)
+        return transformed, counter
+    else:
+        return constraint, counter
```
Can we turn this if/else tree into a registration pattern like in `constraint_generator.py`? It's also kind-of strange to be directly dispatching to another function for half of the constraint types (`CalcConv` etc) but then have logic embedded in the if/else for others

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/constraint_transformation.py:
```diff
@@ -0,0 +1,724 @@
+def transform_conv(constraint, counter):
+    """
+    :param constraint: Calc-conv
+    :return: new counter and the transformed constraint
```
nit: in the PT codebase we use the google docstring style

PT docs: https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#writing-documentation
google docstring style: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

migeed-z replied:
done

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/constraint_transformation.py:
```diff
@@ -0,0 +1,724 @@
+    # return final_result, counter
+
+
+def generate_all_possibilities(my_list):
```
Can you name this something more specific? It's not clear what this does from the name without reading the body of the function

migeed-z replied:
done

jamesr66a reviewed (commented):
This looks fine from a coding perspective, but maybe you could walk us all through what's going on at some point. It would help me (even though I've been working with you on this for a while) and I definitely think it would help @anijain2305  and @jansel

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z commented on torch/fx/experimental/migrate_gradual_types/constraint_transformation.py:
```diff
@@ -0,0 +1,724 @@
+
+
+# run this as the first transformation step till you get a fixed point
+# handles: consistency, matching, precision, GLB, leq, broadcasting, reshape, convolution..
```
done

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z committed:
Update on "first step of constraint transformation"


Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

[ghstack-poisoned]

migeed-z commented:
@pytorchbot merge -g
